lib/Sema/TypeCheckProtocolInference.cpp

```diff
@@ -782,90 +782,59 @@ Type AssociatedTypeInference::computeFixedTypeWitness(
                                             AssociatedTypeDecl *assocType) {
   // Look at all of the inherited protocols to determine whether they
   // require a fixed type for this associated type.
-  Type dependentType = assocType->getDeclaredInterfaceType();
   Type resultType;
   for (auto conformedProto : adoptee->getAnyNominal()->getAllProtocols()) {
     if (!conformedProto->inheritsFrom(assocType->getProtocol()))
       continue;
 
-    auto genericSig = conformedProto->getGenericSignature();
-    if (!genericSig) return Type();
+    if (!conformedProto->getGenericSignature()) {
+      return Type();
+    }
+
+    const auto contextSelf = cast<ArchetypeType>(
+        conformedProto
+            ->mapTypeIntoContext(conformedProto->getSelfInterfaceType())
+            .getPointer());
 
-    Type concreteType = genericSig->getConcreteType(dependentType);
-    if (!concreteType) continue;
+    const auto nestedType =
+        contextSelf->getNestedType(assocType->getName())->mapTypeOutOfContext();
```

Comment:
How about mapTypeIntoContet(assocType->getDeclaredInterfaceType())

Reply:
... oops, I missed the mapTypeOutOfContext(). I think this is easier to accomplish using genericSig->getCanonicalTypeInContext()

Reply:
Currently, we only look through protocols that inherit the protocol of the given associated type, but I was going to remove that limitation eventually, because a fixed type witness can come from an unrelated protocol with a matching associated type: 
```swift
protocol P {
  associatedtype A
}
protocol P2: P where A == Int {}
protocol Q {
  associatedtype A
}
struct Foo: Q, P2 {} // error while checking conformance to Q
```

The problem with `getCanonicalTypeInContext` is that it fails when given a foreign dependent member, so I decided to build on the name of the associated type.

Reply:
What if you use DependentMemberType::get() to get the associated type-less form, using just an identifier? It would still avoid mapping in and out of context.

Reply:
Cool, thanks!